### PR TITLE
[3.11] GH-93516: Speedup line number checks when tracing. (GH-93763)

### PR DESCRIPTION
**Pull request by markshannon**
* Use a lookup table to reduce overhead of getting line numbers during tracing.

Backport of #93763

### CODE REVIEW

**Comment by markshannon** [comment]
The ABI breakage is not really a break, as the new field is inserted (almost) at the end, before any other fields used in C extensions.

File: Tools/scripts/deepfreeze.py [hunk 1]
```diff
@@ -278,6 +279,7 @@ def generate_code(self, name: str, code: types.CodeType) -> str:
             self.write(f".co_name = {co_name},")
             self.write(f".co_qualname = {co_qualname},")
             self.write(f".co_linetable = {co_linetable},")
+            self.write("._co_linearray = NULL,")
```
Comment (Fidget-Spinner) [inline Tools/scripts/deepfreeze.py:282]: Sorry, unrelated question, but does this mean I need to add the field from my `_co_code` change here too?

Comment (markshannon) [reply Tools/scripts/deepfreeze.py:282]: Technically, no. C will initialize any fields not explicitly mentioned to zero.
I think it is good practice to be explicit, though.

**Comment by pablogsal** [comment]
I will regenerate the ABI in this PR

**Comment by markshannon** [comment]
There is no "allow pushes from maintainers" checkbox.
Could you make a PR on this branch?

**Comment by markshannon** [comment]
Closing in favour of https://github.com/python/cpython/pull/94127 to allow edits.

**Pull request closed without merging**